drivers/clock_control/nrf_power_clock.c

```diff
@@ -118,100 +111,69 @@ static const struct nrf_clock_control_sub_config *get_sub_config(
 	return &config->subsys[type];
 }
 
-static enum clock_control_status get_status(struct device *dev,
-					    clock_control_subsys_t subsys)
+static struct onoff_manager *get_onoff_manager(struct device *dev,
+					enum clock_control_nrf_type type)
 {
-	enum clock_control_nrf_type type = (enum clock_control_nrf_type)subsys;
-	struct nrf_clock_control_sub_data *data;
+	struct nrf_clock_control_data *data = dev->driver_data;
 
-	__ASSERT_NO_MSG(type < CLOCK_CONTROL_NRF_TYPE_COUNT);
-	data = get_sub_data(dev, type);
-	if (data->started) {
-		return CLOCK_CONTROL_STATUS_ON;
-	}
+	return &data->mgr[type];
+}
 
-	if (data->ref > 0) {
-		return CLOCK_CONTROL_STATUS_STARTING;
-	}
+DEVICE_DECLARE(clock_nrf);
 
-	return CLOCK_CONTROL_STATUS_OFF;
+struct onoff_manager *z_nrf_clock_control_get_onoff(clock_control_subsys_t sys)
+{
+	return get_onoff_manager(DEVICE_GET(clock_nrf),
+	 			(enum clock_control_nrf_type)sys);
 }
 
-static int clock_stop(struct device *dev, clock_control_subsys_t subsys)
+static enum clock_control_status get_status(struct device *dev,
+					    clock_control_subsys_t subsys)
 {
 	enum clock_control_nrf_type type = (enum clock_control_nrf_type)subsys;
-	const struct nrf_clock_control_sub_config *config;
-	struct nrf_clock_control_sub_data *data;
-	int err = 0;
-	int key;
 
 	__ASSERT_NO_MSG(type < CLOCK_CONTROL_NRF_TYPE_COUNT);
-	config = get_sub_config(dev, type);
-	data = get_sub_data(dev, type);
-
-	key = irq_lock();
-	if (data->ref == 0) {
-		err = -EALREADY;
-		goto out;
-	}
-	data->ref--;
-	if (data->ref == 0) {
-		DBG(dev, subsys, "Stopping");
-		sys_slist_init(&data->list);
-
-		if (IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC_CALIBRATION)
-			&& (subsys == CLOCK_CONTROL_NRF_SUBSYS_LF)) {
-			z_nrf_clock_calibration_lfclk_stopped();
-		}
 
-		nrf_clock_task_trigger(NRF_CLOCK, config->stop_tsk);
+	return GET_STATUS(get_sub_data(dev, type)->flags);
+}
 
-		data->started = false;
+static int set_state(u32_t *flags, u32_t prev_flags, u32_t new_flags)
+{
+	if (atomic_cas(flags, prev_flags, new_flags)) {
+		return 0;
 	}
 
-out:
-	irq_unlock(key);
-
-	return err;
+	return (GET_STATUS(*flags) != GET_STATUS(prev_flags)) ?
+			-EFAULT : -EPERM;
```

Comment:
This seems to be undocumented magic.  Why these two values?  What do they mean?  They propagate to be return values to clock_control_stop(), and require modifications to the generic clock_control API to invoke Nordic-specific recovery, so they're essentially public API, but neither is documented.

Reply:
see https://github.com/zephyrproject-rtos/zephyr/pull/24334#discussion_r452757361